elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java

```diff
@@ -45,53 +56,195 @@
         this.elide = elide;
         this.runner = runner;
         this.asyncQueryDao = asyncQueryDao;
         this.apiVersion = apiVersion;
+        this.resultStorageEngine = resultStorageEngine;
     }
 
    /**
     * This is the main method which processes the Async Query request, executes the query and updates
     * values for AsyncQuery and AsyncQueryResult models accordingly.
     * @return AsyncQueryResult AsyncQueryResult
     * @throws URISyntaxException URISyntaxException
-    * @throws NoHttpResponseException NoHttpResponseException
+    * @throws IOException IOException
     */
-    protected AsyncQueryResult processQuery() throws URISyntaxException, NoHttpResponseException {
+    protected AsyncQueryResult processQuery() throws URISyntaxException, IOException {
         UUID requestId = UUID.fromString(queryObj.getRequestId());
 
         ElideResponse response = null;
+        String responseBody = null;
+        boolean isError = false;
+        Integer recCount = null;
+        Observable<String> observableResult = Observable.empty();
+        QueryType queryType = queryObj.getQueryType();
+        ResultFormatType resultFormatType = queryObj.getResultFormatType();
+        boolean isDownload = queryObj.getResultType() == ResultType.DOWNLOAD;
+
+        if (isDownload && resultStorageEngine == null) {
+            throw new IllegalStateException("ResultStorageEngine unavailable.");
+        }
+
         log.debug("AsyncQuery Object from request: {}", queryObj);
-        if (queryObj.getQueryType().equals(QueryType.JSONAPI_V1_0)) {
+
+        // TODO: Use PersistentResource.loadObjects for Download
+        if (queryType.equals(QueryType.JSONAPI_V1_0)) {
             response = executeJsonApiRequest(queryObj, requestId);
         }
-        else if (queryObj.getQueryType().equals(QueryType.GRAPHQL_V1_0)) {
+        else if (queryType.equals(QueryType.GRAPHQL_V1_0)) {
             response = executeGraphqlRequest(queryObj, requestId);
         }
 
         nullResponseCheck(response);
 
+        responseBody = response.getBody();
+        isError = checkJsonStrErrorMessage(responseBody);
+
         // Create AsyncQueryResult entry for AsyncQuery
         queryResultObj = new AsyncQueryResult();
         queryResultObj.setHttpStatus(response.getResponseCode());
-        queryResultObj.setResponseBody(response.getBody());
-        queryResultObj.setContentLength(response.getBody().length());
+        queryResultObj.setResponseBody(responseBody);
+        queryResultObj.setContentLength(responseBody.length());
         queryResultObj.setCompletedOn(new Date());
 
+        if (isError) {
+            queryResultObj.setResponseBody(responseBody);
+            return queryResultObj;
+        }
+
+        recCount = calculateRecordCount(response, queryType);
+        queryResultObj.setRecordCount(recCount);
+
+        if (resultFormatType == ResultFormatType.CSV) {
+            observableResult = convertJsonToCSV(responseBody);
```

Comment:
This conversion is not save to do in memory.  We do this before checking if the type is download or not.